crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1837 @@
+                    offset,
+                    row_group_data.row_offset + slice_range.end
+                );
+                polars_bail!(ComputeError: msg)
+            };
+
+            // The DataFrame can be empty at this point if no columns were projected from the file,
+            // so we create the row index column manually instead of using `df.with_row_index` to
+            // ensure it has the correct number of rows.
+            let mut ca = IdxCa::from_vec(
+                name,
+                (offset..offset + projection_height as IdxSize).collect(),
+            );
+            ca.set_sorted_flag(IsSorted::Ascending);
+
+            out_columns[0] = ca.into_series();
+        }
+
+        let shared_file_state = row_group_data
+            .shared_file_state
+            .get_or_init(|| async {
+                let path_index = row_group_data.path_index;
+
+                let hive_series = if let Some(hp) = self.hive_partitions.as_deref() {
+                    let mut v = hp[path_index].materialize_partition_columns();
+                    for s in v.iter_mut() {
+                        *s = s.new_from_index(0, row_group_data.file_max_row_group_height);
+                    }
+                    v
+                } else {
+                    vec![]
+                };
+
+                let file_path_series = self.include_file_paths.as_deref().map(|file_path_col| {
+                    StringChunked::full(
+                        file_path_col,
+                        self.paths[path_index].to_str().unwrap(),
+                        row_group_data.file_max_row_group_height,
+                    )
+                    .into_series()
+                });
+
+                SharedFileState {
+                    path_index,
+                    hive_series,
+                    file_path_series,
+                }
+            })
+            .await;
+
+        assert_eq!(shared_file_state.path_index, row_group_data.path_index);
+
+        for s in &shared_file_state.hive_series {
+            debug_assert!(s.len() >= projection_height);
+            out_columns.push(s.slice(0, projection_height));
+        }
+
+        if let Some(file_path_series) = &shared_file_state.file_path_series {
+            debug_assert!(file_path_series.len() >= projection_height);
+            out_columns.push(file_path_series.slice(0, projection_height));
+        }
+
+        let df = unsafe { DataFrame::new_no_checks(out_columns) };
+
+        // Re-calculate: A slice may have been applied.
+        let cols_per_task = 1 + VALUES_PER_THREAD / df.height();
+
+        let df = if let Some(predicate) = self.physical_predicate.as_deref() {
+            let mask = predicate.evaluate_io(&df)?;
+            let mask = mask.bool().unwrap();
+
+            if cols_per_task <= df.width() {
+                df._filter_seq(mask)?
+            } else {
+                let mask = mask.clone();
```

Comment:
Added horizontal parallelism back to predicate filtering here in the row group decoder, but only when we are past `VALUES_PER_THREAD`. I want to do this here so that the predicate is applied before we potentially split the row group into multiple morsels - if we instead do the predicate after then we could end up with very small morsels.